polysemy-plugin/src/Polysemy/Plugin/InlineRecursiveCalls.hs

```diff
@@ -1,91 +1,153 @@
+
+    b -> pure $ (b, Nothing) : pairs
+
+------------------------------------------------------------------------------
+-- | Takes binding group and type of binding inside if there's only one and
+-- returns group with inlining of recursive calls and signatures to be added
+-- to the environment.
+inlineRecCall :: MonadUnique m
+              => ((RecFlag, LHsBinds GhcRn), Maybe (LHsSigWcType GhcRn))
+              -> m ((RecFlag, LHsBinds GhcRn), [LSig GhcRn])
+inlineRecCall (b, mtype)
+  | (Recursive, (bagToList -> [
+        L fun_loc fun_bind@FunBind{ fun_id = L _ fun_name, fun_matches }
+      ])) <- b
+  = do
+      (loopb_name, loopb_decl) <- loopbreaker fun_name
+
+      pure $ case maybeReplaceVarNames fun_name loopb_name fun_matches of
+        Just fun_matches' -> (,)
+          ( Recursive
+          , listToBag
+              [ L fun_loc fun_bind{ fun_matches = fun_matches' }
+              , loopb_decl
+              ]
+          )
+          (  [ inline alwaysInlinePragma                           fun_name
+             , inline defaultInlinePragma{ inl_inline = NoInline } loopb_name
+             ]
+          ++ maybeToList (loopbreakerSig loopb_name <$> mtype)
+          )
+        Nothing -> (b, [])
+    -- We ignore mutually recursive and other bindings
+  | otherwise = pure (b, [])
+
+------------------------------------------------------------------------------
+-- | Creates loopbreaker and it's name from name of original function.
+loopbreaker :: MonadUnique m => Name -> m (Name, LHsBind GhcRn)
+loopbreaker fun_name =
+  (id &&& loopbreakerDecl fun_name) <$> loopbreakerName fun_name
+
+------------------------------------------------------------------------------
+loopbreakerName :: MonadUnique m => Name -> m Name
+loopbreakerName (nameOccName -> occNameFS -> orig_fs) =
+  flip mkSystemVarName (orig_fs <> "__polysemy_loopbreaker") <$> getUniqueM
+
+------------------------------------------------------------------------------
+loopbreakerDecl :: Name -> Name -> LHsBind GhcRn
+loopbreakerDecl fun_name loopb_name =
+  noLoc $ mkTopFunBind Generated (noLoc loopb_name)
+    [ mkSimpleMatch (mkPrefixFunRhs $ noLoc loopb_name) [] $
+        nlHsVar fun_name
+    ]
+
+------------------------------------------------------------------------------
+-- | Creates loopbreaker type signature from type of original function.
+loopbreakerSig :: Name -> LHsSigWcType GhcRn -> LSig GhcRn
+loopbreakerSig loopb_name fun_type =
+  noLoc $ TypeSig NoExt [noLoc loopb_name] fun_type
+
+------------------------------------------------------------------------------
+inline :: InlinePragma -> IdP GhcRn -> LSig GhcRn
+inline how name = noLoc $ InlineSig NoExt (noLoc name) how
+
+------------------------------------------------------------------------------
+-- | Returns value with every variable expression id replaced or 'Nothing' if
+-- there's none to be replaced.
+maybeReplaceVarNames :: Data a => Name -> Name -> a -> Maybe a
```

Comment:
Why combine these two things in one rather than doing a separate pass for checking recursion? Perf worries?

Reply:
Aren't we guaranteed it will be `Just` if the binding group is `Recursive` and only contains one element?

Reply:
Yes
Good point - but should I separate logic for case of one and multiple bindings?

Reply:
Yes
Good point :slightly_smiling_face: